CAT in Tapscript (BIP-347)

arminsabouri opened the pull request:
# CAT in Tapscript

This PR provides the necessary code to enable the opcode OP\_CAT in Tapscript as specified in [BIP-347: OP\_CAT in Tapscript](https://github.com/bitcoin/bips/blob/master/bip-0347.mediawiki) and [BIN-2024-0001](https://github.com/bitcoin-inquisition/binana/blob/master/2024/BIN-2024-0001.md),

**Important:** This PR *does not* include miner activation functionality. This means that merging this PR into bitcoin-core will not make OP\_CAT functional in Bitcoin.
If this PR is merged it is *not* a signal of community consensus around activating CAT nor be read as a portent about the activation process or timeline.
The PR is not a stand-in for consensus around such decisions.

This PR includes:

- An implementation of the Tapscript opcode OP\_CAT along with the flags `SCRIPT_VERIFY_OP_CAT` and `SCRIPT_VERIFY_DISCOURAGE_OP_CAT`.
- Integration and unit tests which ensure that our CAT  implementation works as expected. This includes ensuring that CAT never uses more that 520 bytes of stack memory. A full description of what these tests cover is given below.
- An improvement to `src/test/script_tests.cpp` JSON script format enabling the rapid creation of new Tapscript unit tests.

## Activation on Bitcoin-inquisition (signet)

[Bitcoin-inquisition PR 37](https://github.com/bitcoin-inquisition/bitcoin/pull/39) which contained our implementation of OP\_CAT [was merged into bitcoin-inquisition (signet) on Apr 25 2024](https://github.com/bitcoin-inquisition/bitcoin/pull/39), released as [bitcoin-inquisition release 25.2 on Apr 26 2024.](https://github.com/bitcoin-inquisition/bitcoin/releases/tag/v25.2-inq) and activated in bitcoin-inquisition Apr 30 2024. The bitcoin-inquisition PR was reviewed in the PR review club [(transcript of discussion)](https://bitcoincore.reviews/bitcoin-inquisition-39). Since Apr 30 2024 there have been many OP\_CAT transactions created and spent on signet.

The code merged into bitcoin-inquisition differs from this PR, as we have removed the consensus logic which was used to activate it on signet.

## OP\_CAT Tapscript Implementation

We implement OP\_CAT as a new Tapscript op code by redefining the opcode OP\_SUCCESS126 (126 in decimal and 0x7e in hexadecimal). This is the same opcode value used by the original OP_CAT.

When evaluated, the OP\_CAT instruction:

1. Pops the top two values off the stack,
2. concatenates the popped values together in stack order,
3. and then pushes the concatenated value on the top of the stack.

OP\_CAT fails if there are fewer than two values on the stack or if a concatenated value would have a combined size greater than the maximum script element size of 520 bytes.

See [BIP 347](https://github.com/bitcoin/bips/blob/master/bip-0347.mediawiki) for a deeper description.

### Errors thrown

If evaluated OP\_CAT can throw the following errors:

- If at time of evaluation the stack has fewer than two elements we throw the error: `SCRIPT_ERR_INVALID_STACK_OPERATION`
- If at time of evaluation the top two stack elements have a combined size greater than `MAX_SCRIPT_ELEMENT_SIZE` (520 bytes) we throw the error: `SCRIPT_ERR_PUSH_SIZE`.

### Script verification flags

While this PR does not contain any miner signaling and activation logic and can not activate OP\_CAT, it does contain two flags which future activation logic could set to control activation of OP\_CAT.

- `SCRIPT_VERIFY_OP_CAT` IF a bitcoin node has this set to true, then it treat OP\_CAT enabled for Tapscript. That is, OP\_SUCCESS126 will be redefining to OP\_CAT in Tapscript. If this was set to true at the consensus level this would cause a soft fork.

`SCRIPT_VERIFY_DISCOURAGE_OP_CAT` When set to true, a node receiving any Tapscript transaction containing the opcode OP\_CAT or OP\_SUCCESS126 will reject the transaction throwing the error `SCRIPT_ERR_DISCOURAGE_OP_CAT` but not banning the node which relayed the transaction. This prevents nodes from relaying transactions with OP\_CAT. This  is equivalent to the behavior of `SCRIPT_ERR_DISCOURAGE_OP_CAT = true` when `SCRIPT_VERIFY_OP_CAT = false` as OP\_CAT is an OP\_SUCCESS (OP\_SUCCESS126).

This is how these two flags are intended to be used to ensure a smooth soft fork.

| Stage | `SCRIPT_VERIFY OP_CAT` | `SCRIPT_VERIFY DISCOURAGE_OP_CAT` | Status |
| --- | --- | --- | --- |
| **1.Default** | False | True | This does not represent any change in behavior of the bitcoin-core node. |
| **2.Upgrading** | True | True | OP\_CAT is activating. |
| **3.Upgraded** | True | False | It is clear that OP\_CAT has been activated and the network has been upgraded. |

The flag `SCRIPT_ERR_DISCOURAGE_OP_CAT` provides a window of time for the network to fully activate, before nodes will relay or accept transactions containing OP\_CAT in their mem pools.

## Tests

This PR contains a suite of script tests to ensure that OP\_CAT functions as expected. In the JSON script tests (`script_tests.json`), we test:

- That if OP\_CAT is not activated that there are no changes to bitcoin consensus.
- That regardless of the activation or non-activation of OP\_CAT in Tapscript, pre-Tapscript scripts, i.e. Bitcoin scripts, have no changes in behavior.
- That OP\_CAT when evaluated throws the expected error when there are less than two elements on the stack.
- That OP\_CAT when evaluated in a variety of circumstances and edge cases, successfully concatenates elements of the stack. This includes:
  - multiple calls to OP\_CAT in a row,
  - evaluating OP\_CAT inside of a IF conditional,
  - evaluating OP\_CAT on zero size stack elements, random and large stack elements,
  - evaluating OP\_CAT on values being moved to and from the alt stack,
  - and checking that when evaluated OP\_CAT concatenates the elements in the expected order.

All of these tests are designed to cover the happy path of OP\_CAT, the various errors which OP\_CAT can throw and all the corner cases between those two outcomes.

Additionally we include three tests outside of the JSON script tests.

- `cat_simple` and `cat_empty_stack` are designed to test OP\_CAT outside of the JSON serialization regime. Ensure that we catch bugs that we might miss in the JSON script tests due to a bug introduced at the JSON serialization layer.
- `cat_dup_test` enumerates all stack element sizes from 1 to 522 bytes and then enumerates up to 10 repetitions of `OP\_DUP OP\_CAT`. It then tests if the stack element would exceed 520 bytes and if so did OP\_CAT throw the error `SCRIPT_ERR_PUSH_SIZE`. This allows us to be certain that OP\_CAT will not introduce any `OP\_DUP OP\_CAT` memory exhaustion attacks.

### Better Tapscript tests in JSON script tests

While writing these JSON script tests (`script_tests.json`) we ran into the following problem. The JSON script tests are simple and easy to write for pre-Tapscript scripts, but adding or changing a Tapscript test requires substantial work per test.
Consider the following pre-tapscript test:

```json
["'aa' 'bb'", "CAT 0x4c 0x02 0xaabb EQUAL", "P2SH,STRICTENC", "DISABLED_OPCODE", "CAT disabled"]
```

whereas a Tapscript test for the same script (annotated with comments for better readability) would look like:

```json
[
    [
        "aa",
        "bb",
        "7e4c02aabb87", // output script
        "c0d6889cb081036e0faefa3a35157ad71086b123b2b144b649798b494c300a961d", // control block
        0.00000001
    ],
    "",
    "0x51 0x20 0x15048ed3a65748549c27b671936987093cf73a4c9cb18522a74fb9553060ca99", // Tapscript output
    "P2SH,WITNESS,TAPROOT",
    "OK",
    "TAPSCRIPT CATs aa and bb together and checks if EQUAL to aabb"
]
```

Computing the Tapscript output, such as `0x51 0x20 0x15048ed3a65748549c27b671936987093cf73a4c9cb18522a74fb9553060ca99`, requires writing custom code and running it for each test. The same is true for the Tapscript control block, such as `c0d6889cb081036e0faefa3a35157ad71086b123b2b144b649798b494c300a961d`. If a test is changed or updated new outputs and control blocks must be computed. The complexity of doing this is likely the reason that no one has added any Tapscript tests to JSON script tests until this PR.

In this PR we address this issue by adding the following improvements to JSON script tests:

- Adding simple macros (`"#SCRIPT#` and `#CONTROLBLOCK#`) that allow the script test parser to automatically generate and inject a valid Tapscript output and control block to be computed automatically from the JSON script.
- Allowing Tapscript scripts to use the human readable strings like pre-script scripts by marking the location of the script in the witness stack using `#SCRIPT#`. This transforms the unreadable script `7e4c02aabb87` into `#SCRIPT# CAT 0x4c 0x02 0xaabb EQUAL`.

This results in the following JSON script test which is far easier to write and easier to read.

```json
[
    [
        "aa",
        "bb",
        "#SCRIPT# CAT",
        "#CONTROLBLOCK#",
        0.00000001
    ],
    "",
    "0x51 0x20 #TAPROOTOUTPUT#",
    "P2SH,WITNESS,TAPROOT,OP_CAT",
    "OK",
    "TAPSCRIPT Test of OP_CAT flag by calling CAT on two elements. TAPSCRIPT_OP_CAT flag is set so CAT is executed."
],
```

0xBEEFCAF3 commented:
Drafting until I get the [inquisition PR](https://github.com/bitcoin-inquisition/bitcoin/pull/39) approved and I can get the builds passing.

moonsettler commented:
Some of us have been playing with the idea of neutered CAT: instead of the MAX_SCRIPT_ELEMENT_SIZE (520 bytes) the maximum output size would be 80 bytes.

**Rationale:**
It explicitly disables open ended second order effects by removing the ability to assemble a SIGHASH or CTV template on the stack for more detailed introspection than intended. If such introspection behavior is desired in the future it can be explicitly enabled by specialized and more efficient opcodes.

Combined with CSFS it can be used as a signed datacarrier, for which the 'standard' limit is 80 bytes, it also has to be smaller than 84 bytes which is required for building CTV templates on the stack, and larger than 64/65/72 bytes which are respectively needed for:
- Merkle inclusion: 2x 20/32 byte hashes
- LN-symmetry (LNhance): 2x 32 byte hashes
- Separate sig: 64/72 bytes (0-conf bonds, staking contracts)

Could be a livable compromise between the conservatives that want to preserve certain characteristics of bitcoin and the prometheans who want to give the developers more practical and useful tools to build with.

bigspider commented:
> **Rationale:** It explicitly disables open ended second order effects by removing the ability to assemble a SIGHASH or CTV template on the stack for more detailed introspection than intended. If such introspection behavior is desired in the future it can be explicitly enabled by specialized and more efficient opcodes.

Script is already expressive enough to (awkwardly, expensively) compute arbitrary SHA256 hashes on the stack, except that the result would be broken in smaller pieces of at most 4 bytes. With CAT, those can be concatenated to get a single 32-byte result.

Therefore, neutering CATs does not achieve the desired result of preventing the CHECKSIG tricks, unless you limit the length of the result to less than 32 bytes - which would also neuter most of the utility of the opcode.

arminsabouri committed:
Re-enable OP_CAT in tapscript

Implement OP_CAT as a new Tapscript op code by redefining the opcode OP_SUCCESS126 (126 in decimal and 0x7e in hexadecimal).
This is the same opcode value used by the original OP_CAT.

When evaluated, the OP_CAT instruction:

Pops the top two values off the stack,
concatenates the popped values together in stack order,
and then pushes the concatenated value on the top of the stack.
OP_CAT fails if there are fewer than two values on the stack or if a concatenated value would have a combined size greater than the maximum script element size of 520 bytes.

See [BIP
347](https://github.com/bitcoin/bips/blob/master/bip-0347.mediawiki) for a deeper description.

arminsabouri committed:
test: add unit tests for OP_CAT

Co-authored-by: Ethan Heilman <ethan.r.heilman@gmail.com>

arminsabouri committed:
test: add functional test for OP_CAT spends

The goal of this functional test is to ensure OP_CAT spends are still
disabled by default in segwitv0 and legacy spends. Spending such inputs
should result in `mandatory-script-verify-flag-failed (Attempted to use a disabled opcode)`.
While spending OP_CAT inputs in tapscript should be discouraged under
the default `STANDARD_SCRIPT_VERIFY_FLAGS`.

0xBEEFCAF3 commented on src/script/interpreter.cpp:
```diff
@@ -1800,10 +1819,19 @@ static bool ExecuteWitnessScript(const Span<const valtype>& stack_span, const CS
-                if (flags & SCRIPT_VERIFY_DISCOURAGE_OP_SUCCESS) {
-                    return set_error(serror, SCRIPT_ERR_DISCOURAGE_OP_SUCCESS);
+                if (opcode == OP_CAT) {
+                    if (flags & SCRIPT_VERIFY_DISCOURAGE_OP_CAT) {
```
Great suggestion! Added some additional unit tests to script_tests.json testing different combinations of scripts with and without (SCRIPT_VERIFY_DISCOURAGE_OP_CAT, SCRIPT_VERIFY_OP_CAT and SCRIPT_VERIFY_DISCOURAGE_OP_SUCCESS)

ajtowns replied:
Added to the inquisition PR in https://github.com/bitcoin-inquisition/bitcoin/pull/68

achow101 commented:
Converting to draft while waiting for consensus on this.

EthanHeilman commented:
@achow101 What does draft vs non-draft mean in the bitcoin-core github? My understanding was moving from draft to non-draft means code is ready for review. Similar to moving WIP.

Is draft/non-draft is being used to signal community consensus or bitcoin-core consensus? If so, I agree this should be a draft.

achow101 commented:
In general, draft means that a PR is not in a state where it could be merged. It indicates to reviewers that they may not want to do any in depth code review of the PR yet.

In the context of this PR, there does not appear to be consensus for the concept of OP_CAT in TapScript yet, so this PR cannot be merged even if the code is okay. Hence marking it as a draft.

Note that this is a result of a [Kill/Shill/Merge](https://github.com/bitcoin-core/bitcoin-devwiki/wiki/Kill-Shill-Drill) session that occurred at the most recent CoreDev.

EthanHeilman commented:
@achow101 Thanks, that's helpful for understanding the context. If consensus forms around OP_CAT should I or @0xBEEFCAF3 mark this as no longer as draft? I can't speak for @0xBEEFCAF3 but I would feel uncomfortable making that decision as it would open me up to criticism that I was attempting to manufacture consensus, even if there was consensus. What is the process here?

achow101 commented:
> If consensus forms around OP_CAT should I or @0xBEEFCAF3 mark this as no longer as draft?

Yes. A maintainer may also do so as well.

AlexSQY commented:
What about SCRIPT_VERIFY OP_CAT = False and SCRIPT_VERIFY DISCOURAGE_OP_CAT = False?
This combination is not described in the "soft fork" table. Is it part of the tests?

EthanHeilman commented:
@AlexSQY That's an interesting question, we are you concerned about here?

Current Bitcoin behavior is `OP_CAT = False` and `DISCOURAGE_OP_CAT = False`. I don't believe we would ever want to set  `DISCOURAGE_OP_CAT = False` while `OP_CAT = False` since a disabled opcode should be discouraged.

EthanHeilman commented:
@AlexSQY I see your point. Let me see if there is anything that can be done to ensure it is never the case that `OP_CAT = False && DISCOURAGE_OP_CAT = False` since that should never happen outside of tests. 

The behavior is what you see in that test is the expected behavior for `DISCOURAGE_OP_CAT = False` and `OP_CAT = False`,
since `OP_CAT = False` the CAT opcode is an OP_SUCCESS so it returns success (`OK`).

@0xBEEFCAF3 What do you think?

0xBEEFCAF3 commented:
>Let me see if there is anything that can be done to ensure it is never the case that OP_CAT = False && DISCOURAGE_OP_CAT = False since that should never happen outside of tests.

By default, the `SCRIPT_VERIFY_DISCOURAGE_OP_CAT` bit [is set](https://github.com/bitcoin/bitcoin/pull/29247/files#diff-1fc0f6b5081e8ed5dfa8bf230744ad08cc6f4c1147e98552f1f424b0492fe9bdR121), and we do not expose this as a configurable mempool policy. The only practical way for node operators to unset this bit (`SCRIPT_VERIFY_DISCOURAGE_OP_CAT = False`) is by compiling a custom binary.

That said, @AlexSQY, you're right -- this edge case deserves documentation. If I recall correctly, it’s being tested [here](https://github.com/bitcoin/bitcoin/pull/29247/files#diff-ef0159635ee0fee8864cd9570a04337cabb63e9c49a9e96d7b746194ecaa427dR2541), where neither the discourage bit nor the script verify bit is set. In that case, both are `false`, and `0x7e` just behaves as `OP_SUCCESS`.

ajtowns commented:
> Current Bitcoin behavior is `OP_CAT = False` and `DISCOURAGE_OP_CAT = False`.

DISCOURAGE_OP_SUCCESS in current bitcoin covers the behaviour of DISCOURAGE_OP_CAT.

OP_CAT=false and DISCOURAGE_OP_CAT=false are used for validating blocks prior to activation of OP_CAT, so this isn't an edge case as far as I can see? OP_CAT=true and DISCOURAGE_OP_CAT=true should be an edge case, but it doesn't make sense of course: "disallow OP_CAT, but if it were allowed, which it's not, enforce its rules".

I think the table should look more like:

| Phase  | Mode | SV_OP_CAT | SV_DISCOURAGE_OP_CAT |
| ---- | ---- | ---- | ---- |
| Pre-activation | Transaction relay | :x: (false) | :heavy_check_mark: (true) |
| Pre-activation | Block validation | :x:  (false) | :x:  (false) |
| Post-activation | Transaction relay | :heavy_check_mark: (true) | :x:  (false) |
| Post-activation | Block validation | :heavy_check_mark: (true) | :x:  (false) |

EthanHeilman commented:
@DrahtBot I am still very interested in merging this, however there was unexpected progress in quantum computers and I am temporarily shifting my attention to that. I plan to return this after helping with quantum (hopefully in 2026).
